README.md
[Repository](https://github.com/pallet/graphite-crate) &#xb7;
[Issues](https://github.com/pallet/graphite-crate/issues) &#xb7;
[API docs](http://palletops.com/graphite-crate/0.8/api) &#xb7;
[Annotated source](http://palletops.com/graphite-crate/0.8/annotated/uberdoc.html) &#xb7;
[Release Notes](https://github.com/pallet/graphite-crate/blob/develop/ReleaseNotes.md)

Install and configure graphite.

### Dependency Information

```clj
:dependencies [[com.palletops/graphite-crate "0.8.0-alpha.1"]]
```

### Releases

<table>
<thead>
  <tr><th>Pallet</th><th>Crate Version</th><th>Repo</th><th>GroupId</th></tr>
</thead>
<tbody>
  <tr>
    <th>0.8.0-beta.8</th>
    <td>0.8.0-alpha.1</td>
    <td>clojars</td>
    <td>com.palletops</td>
    <td><a href='https://github.com/pallet/graphite-crate/blob/0.8.0-alpha.1/ReleaseNotes.md'>Release Notes</a></td>
    <td><a href='https://github.com/pallet/graphite-crate/blob/0.8.0-alpha.1/'>Source</a></td>
  </tr>
</tbody>
</table>

The `server-spec` function provides a server spec you can add to an `:extends`
clause.

# graphite-crate

A crate to install and configure graphite.

## Usage

Add `[com.palletops/graphite-crate "0.1.0"]` to the `:dependencies` in your
`project.clj` file.


## License

Licensed under [EPL](http://www.eclipse.org/legal/epl-v10.1.1.html)

Copyright © 2012-2014 Hugo Duncan, Antoni Batchelli



ReleaseNotes.md
## 0.8.0-alpha.3

- Update pallet and clojure dependency versions

## 0.8.0-alpha.2

- Remove live-test in preparation of using test-env.

- Update README with new release notice, copyright ownership.

- Initialize for release automation via lein-pallet-release.

- Prepare for lein-pallet-release.

- Update license to EPL (from All Rights Reserved).

- Fix USAGE.md

## 0.8.0-alpha.1

- Update to latest pallet and crate conventions

## 0.1.0

Initial release.


test/pallet/crate/graphite_test.clj
(ns pallet.crate.graphite-test
  (:require
   [clojure.test :refer :all]
   [pallet.crate.graphite :as graphite]
   [pallet.crate.network-service :refer [wait-for-port-listen]]
   [pallet.actions :refer [package-manager]]
   [pallet.algo.fsmop :refer [complete?]]
   [pallet.api :refer [lift plan-fn group-spec server-spec]]
   [pallet.crate.automated-admin-user :refer [automated-admin-user]]
   #_[pallet.live-test :refer [images test-nodes]]))

(def graphite-test-spec
  (group-spec "graphite"
    :count 1
    :extends [(graphite/server-spec {})]
    :phases {:bootstrap (plan-fn
                          (automated-admin-user)
                          (package-manager :update))
             :test (plan-fn
                     (wait-for-port-listen 8080))}
    :roles #{:live-test}))

#_(deftest ^:live-test live-test

  (doseq [image (images)]
    (test-nodes
        [compute node-map node-types [:install :configure]]
      {:graphite
       (assoc graphite-test-spec :image image)})))


release.sh
#!/bin/bash

# release collectd

if [[ $# -lt 3 ]]; then
  echo "usage: $(basename $0) previous-version new-version next-version" >&2
  exit 1
fi

previous_version=$1
version=$2
next_version=$3

echo ""
echo "Start release of $version, previous version is $previous_version"
echo ""
echo ""

lein do clean, with-profile +no-checkouts test && \
git flow release start $version || exit 1

lein with-profile release set-version ${version} :previous-version ${previous_version} \
  || { echo "set version failed" >2 ; exit 1; }

echo ""
echo ""
echo "Changes since $previous_version"
git --no-pager log --pretty=changelog $previous_version..
echo ""
echo ""
echo "Now edit project.clj, ReleaseNotes and README"

$EDITOR resources/pallet_crate/graphite_crate/meta.edn
lein crate-doc
$EDITOR project.clj
$EDITOR ReleaseNotes.md

echo -n "commiting project.clj, release notes and readme.  enter to continue:" \
&& read x \
&& git add project.clj ReleaseNotes.md README.md resources/pallet_crate/graphite_crate/meta.edn \
&& git commit -m "Updated project.clj, release notes and readme for $version" \
&& echo -n "Peform release.  enter to continue:" && read x \
&& lein do clean, with-profile +no-checkouts test, with-profile +palletops deploy palletops \
&& git flow release finish $version \
&& echo "Now push to github. Don't forget the tags!" \
&& lein with-profile release set-version ${next_version} \
&& git add project.clj \
&& git commit -m "Updated version for next release cycle"


src/pallet/crate/graphite/config.clj
(ns pallet.crate.graphite.config)

(def carbon-config
  {:cache
   {
    ;; Configure carbon directories.
    ;;
    ;; OS environment variables can be used to tell carbon where graphite is
    ;; installed, where to read configuration from and where to write data.
    ;;
    ;;   GRAPHITE_ROOT        - Root directory of the graphite installation.
    ;;                          Defaults to ../
    ;;   GRAPHITE_CONF_DIR    - Configuration directory (where this file lives).
    ;;                          Defaults to $GRAPHITE_ROOT/conf/
    ;;   GRAPHITE_STORAGE_DIR - Storage directory for whipser/rrd/log/pid files.
    ;;                          Defaults to $GRAPHITE_ROOT/storage/
    ;;
    ;; To change other directory paths, add settings to this file. The following
    ;; configuration variables are available with these default values:
    ;;
    ;;   STORAGE_DIR    = $GRAPHITE_STORAGE_DIR
    ;;   LOCAL_DATA_DIR = STORAGE_DIR/whisper/
    ;;   WHITELISTS_DIR = STORAGE_DIR/lists/
    ;;   CONF_DIR       = STORAGE_DIR/conf/
    ;;   LOG_DIR        = STORAGE_DIR/log/
    ;;   PID_DIR        = STORAGE_DIR/
    ;;
    ;; For FHS style directory structures, use:
    ;;
    ;;   STORAGE_DIR    = /var/lib/carbon/
    ;;   CONF_DIR       = /etc/carbon/
    ;;   LOG_DIR        = /var/log/carbon/
    ;;   PID_DIR        = /var/run/
    ;;
    ;;LOCAL_DATA_DIR = /opt/graphite/storage/whisper/

    ;; Specify the user to drop privileges to
    ;; If this is blank carbon runs as the user that invokes it
    ;; This user must have write access to the local data directory
    :USER ""

    ;; Limit the size of the cache to avoid swapping or becoming CPU bound.
    ;; Sorts and serving cache queries gets more expensive as the cache grows.
    ;; Use the value "inf" (infinity) for an unlimited cache size.
    :MAX_CACHE_SIZE "inf"

    ;; Limits the number of whisper update_many() calls per second, which effectively
    ;; means the number of write requests sent to the disk. This is intended to
    ;; prevent over-utilizing the disk and thus starving the rest of the system.
    ;; When the rate of required updates exceeds this, then carbon's caching will
    ;; take effect and increase the overall throughput accordingly.
    :MAX_UPDATES_PER_SECOND 500

    ;; Softly limits the number of whisper files that get created each minute.
    ;; Setting this value low (like at 50) is a good way to ensure your graphite
    ;; system will not be adversely impacted when a bunch of new metrics are
    ;; sent to it. The trade off is that it will take much longer for those metrics'
    ;; database files to all get created and thus longer until the data becomes usable.
    ;; Setting this value high (like "inf" for infinity) will cause graphite to create
    ;; the files quickly but at the risk of slowing I/O down considerably for a while.
    :MAX_CREATES_PER_MINUTE 50

    :LINE_RECEIVER_INTERFACE "0.0.0.0"
    :LINE_RECEIVER_PORT 2003

    ;; Set this to True to enable the UDP listener. By default this is off
    ;; because it is very common to run multiple carbon daemons and managing
    ;; another (rarely used) port for every carbon instance is not fun.
    :ENABLE_UDP_LISTENER "False"
    :UDP_RECEIVER_INTERFACE "0.0.0.0"
    :UDP_RECEIVER_PORT 2003

    :PICKLE_RECEIVER_INTERFACE "0.0.0.0"
    :PICKLE_RECEIVER_PORT 2004

    ;; Per security concerns outlined in Bug ;;817247 the pickle receiver
    ;; will use a more secure and slightly less efficient unpickler.
    ;; Set this to True to revert to the old-fashioned insecure unpickler.
    :USE_INSECURE_UNPICKLER "False"

    :CACHE_QUERY_INTERFACE "0.0.0.0"
    :CACHE_QUERY_PORT 7002

    ;; Set this to False to drop datapoints received after the cache
    ;; reaches MAX_CACHE_SIZE. If this is True (the default) then sockets
    ;; over which metrics are received will temporarily stop accepting
    ;; data until the cache size falls below 95% MAX_CACHE_SIZE.
    :USE_FLOW_CONTROL "True"

    ;; By default, carbon-cache will log every whisper update. This can be excessive and
    ;; degrade performance if logging on the same volume as the whisper data is stored.
    :LOG_UPDATES "False"

    ;; On some systems it is desirable for whisper to write synchronously.
    ;; Set this option to True if you'd like to try this. Basically it will
    ;; shift the onus of buffering writes from the kernel into carbon's cache.
    :WHISPER_AUTOFLUSH "False"

    ;; By default new Whisper files are created pre-allocated with the data region
    ;; filled with zeros to prevent fragmentation and speed up contiguous reads and
    ;; writes (which are common). Enabling this option will cause Whisper to create
    ;; the file sparsely instead. Enabling this option may allow a large increase of
    ;; MAX_CREATES_PER_MINUTE but may have longer term performance implications
    ;; depending on the underlying storage configuration.
    ;; WHISPER_SPARSE_CREATE "False"

    ;; Enabling this option will cause Whisper to lock each Whisper file it writes
    ;; to with an exclusive lock (LOCK_EX, see: man 2 flock). This is useful when
    ;; multiple carbon-cache daemons are writing to the same files
    ;; WHISPER_LOCK_WRITES "False"

    ;; Set this to True to enable whitelisting and blacklisting of metrics in
    ;; CONF_DIR/whitelist and CONF_DIR/blacklist. If the whitelist is missing or
    ;; empty, all metrics will pass through
    ;; USE_WHITELIST "False"

    ;; By default, carbon itself will log statistics (such as a count,
    ;; metricsReceived) with the top level prefix of 'carbon' at an interval of 60
    ;; seconds. Set CARBON_METRIC_INTERVAL to 0 to disable instrumentation
    ;; CARBON_METRIC_PREFIX carbon
    ;; CARBON_METRIC_INTERVAL 60

    ;; Enable AMQP if you want to receve metrics using an amqp broker
    ;; ENABLE_AMQP "False"

    ;; Verbose means a line will be logged for every metric received
    ;; useful for testing
    ;; AMQP_VERBOSE "False"

    ;; AMQP_HOST localhost
    ;; AMQP_PORT 5672
    ;; AMQP_VHOST /
    ;; AMQP_USER guest
    ;; AMQP_PASSWORD guest
    ;; AMQP_EXCHANGE graphite
    ;; AMQP_METRIC_NAME_IN_BODY "False"

    ;; The manhole interface allows you to SSH into the carbon daemon
    ;; and get a python interpreter. BE CAREFUL WITH THIS! If you do
    ;; something like time.sleep() in the interpreter, the whole process
    ;; will sleep! This is *extremely* helpful in debugging, assuming
    ;; you are familiar with the code. If you are not, please don't
    ;; mess with this, you are asking for trouble :)
    ;;
    ;; ENABLE_MANHOLE "False"
    ;; MANHOLE_INTERFACE 127.0.0.1
    ;; MANHOLE_PORT 7222
    ;; MANHOLE_USER admin
    ;; MANHOLE_PUBLIC_KEY ssh-rsa AAAAB3NzaC1yc2EAAAABiwAaAIEAoxN0sv/e4eZCPpi3N3KYvyzRaBaMeS2RsOQ/cDuKv11dlNzVeiyc3RFmCv5Rjwn/lQ79y0zyHxw67qLyhQ/kDzINc4cY41ivuQXm2tPmgvexdrBv5nsfEpjs3gLZfJnyvlcVyWK/lId8WUvEWSWHTzsbtmXAF2raJMdgLTbQ8wE=

    ;; Patterns for all of the metrics this machine will store. Read more at
    ;; http://en.wikipedia.org/wiki/Advanced_Message_Queuing_Protocol;;Bindings
    ;;
    ;; Example: store all sales, linux servers, and utilization metrics
    ;; BIND_PATTERNS sales.;;, servers.linux.;;, ;;.utilization
    ;;
    ;; Example: store everything
    ;; BIND_PATTERNS ;;
    }
   ;; To configure special settings for the carbon-cache instance 'b', uncomment this:
   ;;[cache:b]
   ;;LINE_RECEIVER_PORT 2103
   ;;PICKLE_RECEIVER_PORT 2104
   ;;CACHE_QUERY_PORT 7102
   ;; and any other settings you want to customize, defaults are inherited
   ;; from [carbon] section.
   ;; You can then specify the --instance=b option to manage this instance

   :relay {
    :LINE_RECEIVER_INTERFACE "0.0.0.0"
    :LINE_RECEIVER_PORT 2013
    :PICKLE_RECEIVER_INTERFACE "0.0.0.0"
    :PICKLE_RECEIVER_PORT 2014

    ;; To use consistent hashing instead of the user defined relay-rules.conf,
    ;; change this to:
    ;; RELAY_METHOD consistent-hashing
    :RELAY_METHOD "rules"

    ;; If you use consistent-hashing you may want to add redundancy
    ;; of your data by replicating every datapoint to more than
    ;; one machine.
    :REPLICATION_FACTOR 1

    ;; This is a list of carbon daemons we will send any relayed or
    ;; generated metrics to. The default provided would send to a single
    ;; carbon-cache instance on the default port. However if you
    ;; use multiple carbon-cache instances then it would look like this:
    ;;
    ;; DESTINATIONS 127.0.0.1:2004:a, 127.0.0.1:2104:b
    ;;
    ;; The general form is IP:PORT:INSTANCE where the :INSTANCE part is
    ;; optional and refers to the "None" instance if omitted.
    ;;
    ;; Note that if the destinations are all carbon-caches then this should
    ;; exactly match the webapp's CARBONLINK_HOSTS setting in terms of
    ;; instances listed (order matters!).
    ;;
    ;; If using RELAY_METHOD rules, all destinations used in relay-rules.conf
    ;; must be defined in this list
    :DESTINATIONS "127.0.0.1:2004"

    ;; This defines the maximum "message size" between carbon daemons.
    ;; You shouldn't need to tune this unless you really know what you're doing.
    :MAX_DATAPOINTS_PER_MESSAGE 500
    :MAX_QUEUE_SIZE 10000

    ;; Set this to False to drop datapoints when any send queue (sending datapoints
    ;; to a downstream carbon daemon) hits MAX_QUEUE_SIZE. If this is True (the
    ;; default) then sockets over which metrics are received will temporarily stop accepting
    ;; data until the send queues fall below 80% MAX_QUEUE_SIZE.
    :USE_FLOW_CONTROL "True"

    ;; Set this to True to enable whitelisting and blacklisting of metrics in
    ;; CONF_DIR/whitelist and CONF_DIR/blacklist. If the whitelist is missing or
    ;; empty, all metrics will pass through
    ;; USE_WHITELIST "False"

    ;; By default, carbon itself will log statistics (such as a count,
    ;; metricsReceived) with the top level prefix of 'carbon' at an interval of 60
    ;; seconds. Set CARBON_METRIC_INTERVAL to 0 to disable instrumentation
    ;; CARBON_METRIC_PREFIX carbon
    ;; CARBON_METRIC_INTERVAL 60
    }

   :aggregator {
    :LINE_RECEIVER_INTERFACE "0.0.0.0"
    :LINE_RECEIVER_PORT 2023

    :PICKLE_RECEIVER_INTERFACE "0.0.0.0"
    :PICKLE_RECEIVER_PORT 2024

    ;; This is a list of carbon daemons we will send any relayed or
    ;; generated metrics to. The default provided would send to a single
    ;; carbon-cache instance on the default port. However if you
    ;; use multiple carbon-cache instances then it would look like this:
    ;;
    ;; DESTINATIONS 127.0.0.1:2004:a, 127.0.0.1:2104:b
    ;;
    ;; The format is comma-delimited IP:PORT:INSTANCE where the :INSTANCE part is
    ;; optional and refers to the "None" instance if omitted.
    ;;
    ;; Note that if the destinations are all carbon-caches then this should
    ;; exactly match the webapp's CARBONLINK_HOSTS setting in terms of
    ;; instances listed (order matters!).
    :DESTINATIONS "127.0.0.1:2004"

    ;; If you want to add redundancy to your data by replicating every
    ;; datapoint to more than one machine, increase this.
    :REPLICATION_FACTOR 1

    ;; This is the maximum number of datapoints that can be queued up
    ;; for a single destination. Once this limit is hit, we will
    ;; stop accepting new data if USE_FLOW_CONTROL is True, otherwise
    ;; we will drop any subsequently received datapoints.
    :MAX_QUEUE_SIZE 10000

    ;; Set this to False to drop datapoints when any send queue (sending datapoints
    ;; to a downstream carbon daemon) hits MAX_QUEUE_SIZE. If this is True (the
    ;; default) then sockets over which metrics are received will temporarily stop accepting
    ;; data until the send queues fall below 80% MAX_QUEUE_SIZE.
    :USE_FLOW_CONTROL "True"

    ;; This defines the maximum "message size" between carbon daemons.
    ;; You shouldn't need to tune this unless you really know what you're doing.
    :MAX_DATAPOINTS_PER_MESSAGE 500

    ;; This defines how many datapoints the aggregator remembers for
    ;; each metric. Aggregation only happens for datapoints that fall in
    ;; the past MAX_AGGREGATION_INTERVALS * intervalSize seconds.
    :MAX_AGGREGATION_INTERVALS 5

    ;; Set this to True to enable whitelisting and blacklisting of metrics in
    ;; CONF_DIR/whitelist and CONF_DIR/blacklist. If the whitelist is missing or
    ;; empty, all metrics will pass through
    ;; USE_WHITELIST "False"

    ;; By default, carbon itself will log statistics (such as a count,
    ;; metricsReceived) with the top level prefix of 'carbon' at an interval of 60
    ;; seconds. Set CARBON_METRIC_INTERVAL to 0 to disable instrumentation
    ;; CARBON_METRIC_PREFIX carbon
    ;; CARBON_METRIC_INTERVAL 60
    }})


src/pallet/crate/graphite.clj
;;; Copyright 2012, 2013 Hugo Duncan.
;;; All rights reserved.

;; http://geek.michaelgrace.org/2011/09/how-to-install-graphite-on-ubuntu/
;; http://linuxracker.com/2012/03/31/setting-up-graphite-server-on-debian-squeeze/
;; http://stackoverflow.com/questions/11491268/install-pycairo-in-virtualenv

(ns pallet.crate.graphite
  "A pallet crate to install and configure graphite"
  (:require
   [clojure.string :refer [join]]
   [pallet.action :refer [with-action-options]]
   [pallet.actions
    :refer [assoc-settings directory exec-checked-script exec-script group
            packages remote-directory remote-file sed service symbolic-link
            user]
    :rename {user user-action group group-action
             assoc-settings assoc-settings-action
             service service-action}]
   [pallet.api :refer [plan-fn] :as api]
   [pallet.config-file.format :refer [sectioned-properties]]
   [pallet.crate
    :refer [defplan assoc-settings defmethod-plan get-settings
            get-node-settings group-name nodes-with-role target-id]]
   [pallet.crate-install :as crate-install]
   [pallet.crate.graphite.config :refer [carbon-config]]
   [pallet.script.lib :refer [pid-root log-root config-root user-home]]
   [pallet.stevedore :refer [fragment script]]
   [pallet.utils :refer [apply-map]]
   [pallet.version-dispatch
    :refer [defmulti-version-plan defmethod-version-plan]]))


(def ^{:doc "Flag for recognising changes to configuration"}
  graphite-config-changed-flag "graphite-config")

;;; # Settings
(defn default-settings []
  {:version "0.9.10"
   :user "graphite"
   :owner "graphite"
   :group "graphite"
   :plugin-dir "/var/lib/graphite"
   :home "/opt/graphite"
   :config-dir "/opt/graphite/conf"
   :carbon-config carbon-config
   :storage-schemas (array-map
                     :carbon {:pattern "^carbon\\."
                              :retentions "60:90d"}
                     :default_1min_for_1day {:pattern ".*"
                                             :retentions "60s:1d"})
   :storage-aggregation {}
   :relay-rules nil
   :aggregation-rules nil
   :webapp-bind-address "127.0.0.0:8080"
   :service "graphite"})

;;; At the moment we just have a single implementation of settings,
;;; but this is open-coded.
(defmulti-version-plan settings-map [version settings])

(defmethod-version-plan
    settings-map {:os :linux}
    [os os-version version settings]
  (cond
   (:install-strategy settings) settings
   :else (assoc settings :install-strategy ::virtualenv)))

(defplan settings
  "Settings for graphite"
  [{:keys [user owner group dist dist-urls cloudera-version version
           instance-id]
    :as settings}]
  (let [settings (merge (default-settings) settings)
        settings (settings-map (:version settings) settings)]
    (assert (:install-strategy settings) "No install strategy specified.")
    (assoc-settings :graphite settings {:instance-id instance-id})))

;;; # Install
(defmethod-plan crate-install/install ::virtualenv
  [facility instance-id]
  (let [{:keys [user owner group src-dir home url] :as settings}
        (get-settings facility {:instance-id instance-id})]
    (packages
     :yum ["python-virtualenv" "python-pip" "python-dev"]
     :apt ["python-virtualenv" "python-setuptools" "python-dev"
           "python-cairo"]
     :aptitude ["python-virtualenv" "python-setuptools" "python-dev"
                "python-cairo"])
    (directory home :owner user :group group)
    (exec-checked-script
     "Allow pallet script execution as other users"
     ("chmod" "711" .))
    (with-action-options
      {:sudo-user user :script-dir (fragment (~user-home ~user))}
      (exec-checked-script
       "Install graphite with virtualenv"
       (when (not (file-exists? (str ~home "/bin/python")))
         ("virtualenv" ~home))
       ("source" (str ~home "/bin/activate"))
       ("export" (set! HOME ~home))
       ("pip" -q install whisper)
       ("pip" -q install carbon)
       ("pip" -q install django)
       ("pip" -q install django-tagging)
       ("pip" -q install graphite-web)
       ("easy_install" gunicorn)
       ;; TODO
       ;; edit local_settings.py to enable DATABASE
       ))
    (remote-file (str home "/lib/python2.7/site-packages/cairo")
                 :link "/usr/lib/python2.7/dist-packages/cairo")))

(defplan install
  "Install graphite."
  [{:keys [instance-id]}]
  (let [settings (get-settings :graphite {:instance-id instance-id})]
    (crate-install/install :graphite instance-id)))


;; (defplan install-gunicorn
;;   "Install gunicorn"
;;   [{:keys [instance-id] :as options}]
;;   [{:keys [home service user]}
;;    (get-settings :graphite {:instance-id instance-id})]
;;   (with-action-options {:script-dir home :sudo-user user}
;;     (exec-checked-script
;;      (str "Graphite " daemon " " action)
;;      ("easy_install" gunicorn))))

;;; # User
(defplan user
  "Create the graphite user"
  [{:keys [instance-id] :as options}]
  (let [{:keys [user owner group home]} (get-settings :graphite options)]
    (group-action group :system true)
    (when (not= owner user)
      (user-action owner :group group :system true))
    (user-action
     user :group group :system true :create-home true :shell :bash)))


;;; # Configuration
(defplan config-file
  "Helper to write config files"
  [{:keys [owner group config-dir] :as settings} filename file-source]
  (apply
   remote-file (str config-dir "/" filename)
   :flag-on-changed graphite-config-changed-flag
   :owner owner :group group
   (apply concat file-source)))

(defplan configure
  "Write all config files"
  [{:keys [instance-id] :as options}]
  (let [{:keys [carbon-config storage-aggregation storage-schemas home user
                owner group]
         :as settings}
        (get-settings :graphite options)]
    (config-file settings "carbon.conf"
                 {:content (sectioned-properties carbon-config)})
    (config-file settings "storage-schemas.conf"
                 {:content (sectioned-properties storage-schemas)})
    (config-file settings "storage-aggregation.conf"
                 {:content (sectioned-properties storage-aggregation)})
    (remote-file
     (str home "/webapp/graphite/local_settings.py")
     :content
     "DATABASES = {
    'default': {
        'NAME': '/opt/graphite/storage/graphite.db',
        'ENGINE': 'django.db.backends.sqlite3',
        'USER': '',
        'PASSWORD': '',
        'HOST': '',
        'PORT': ''
    }
}"
     :owner owner :group group)
    (sed (str home "/webapp/graphite/app_settings.py")
         {"SECRET_KEY = ''" "SECRET_KEY = 'abcdef'"})
    (with-action-options {:sudo-user user
                          :script-dir (str home "/webapp/graphite/")}
      (exec-checked-script
       "Initialise database"
       ("source" (str ~home "/bin/activate"))
       ("python" manage.py syncdb "--noinput")))))

(defplan service
  "Control a graphite service.

   Specify `:if-config-changed true` to make actions conditional on a change in
   configuration.

   Other options are as for `pallet.action.service/service`. The service
   name is looked up in the request parameters."
  [daemon {:keys [action if-config-changed if-flag instance-id]
           :or {action :start}
           :as options}]
  (let [{:keys [home service user]}
        (get-settings :graphite {:instance-id instance-id})
        options (if if-config-changed
                  (assoc options :if-flag graphite-config-changed-flag)
                  options)]
    (with-action-options {:script-dir home :sudo-user user}
      (exec-checked-script
       (str "Graphite " daemon " " action)
       ("source" (str ~home "/bin/activate"))
       (if (not ((str ~home "/bin/" ~daemon ".py") status))
         ("nohup" (str ~home "/bin/" ~daemon ".py") ~(name action)))))))

(defplan web-server
  "Start a graphite web service"
  [{:keys [instance-id] :as options}]
  (let [{:keys [home service user group webapp-bind-address]}
        (get-settings :graphite {:instance-id instance-id})]
    (with-action-options {:script-dir home}
      (exec-script
       ;; TODO work out how to check if this is already running
       ;; (str "Graphite web server")
       ("source" (str ~home "/bin/activate"))
       (when (not (pipe ("ps" ax) ("grep" gunicorn_django) ("grep" -v grep)))
         ("("
          ("nohup"
           (str ~home "/bin/gunicorn_django")
           -u ~user -g ~group -b ~webapp-bind-address
           (str "--access-logfile=" ~home "/storage/log/webapp/gunicorn.log")
           (str "--pythonpath=" ~home "/webapp")
           "graphite.settings")
          "& )")
         ("sleep" 5))))))                 ; allow sub-process to start

(defn server-spec
  "Returns a server-spec that installs and configures graphite"
  [settings & {:keys [instance-id] :as options}]
  (api/server-spec
   :phases
   {:settings (plan-fn
                (pallet.crate.graphite/settings (merge settings options)))
    :install (plan-fn
               (user options)
               (install options))
    :configure (plan-fn
                 (configure options)
                 (service "carbon-cache" options)
                 (web-server options))
    :start (plan-fn
             (service "carbon-cache" options)
             (web-server options))}))
